File: clients/python/src/graphdb_client/aio/resources/nodes.py
from __future__ import annotations

from typing import Any, AsyncIterator, Mapping, Sequence

from ...models import Node
from ..transport import AsyncTransport


class AsyncNodesResource:
    def __init__(self, transport: AsyncTransport) -> None:
        self._t = transport

    async def create(
        self, labels: Sequence[str], properties: Mapping[str, Any] | None = None
    ) -> Node:
        res = await self._t.request(
            "POST", "/nodes",
            json={"labels": list(labels), "properties": dict(properties or {})},
        )
        return Node.from_dict(res.data)

    async def get(self, node_id: int) -> Node:
        res = await self._t.request("GET", f"/nodes/{node_id}")
        return Node.from_dict(res.data)

    async def update(self, node_id: int, properties: Mapping[str, Any]) -> Node:
        res = await self._t.request(
            "PUT", f"/nodes/{node_id}", json={"properties": dict(properties)}
        )
        return Node.from_dict(res.data)

    async def delete(self, node_id: int) -> None:
        await self._t.request("DELETE", f"/nodes/{node_id}")

    async def batch_create(self, nodes: Sequence[Mapping[str, Any]]) -> list[Node]:
        payload = {"nodes": [
            {"labels": list(n.get("labels", [])), "properties": dict(n.get("properties", {}))}
            for n in nodes
        ]}
        res = await self._t.request("POST", "/nodes/batch", json=payload)
        return [Node.from_dict(d) for d in (res.data.get("nodes") or [])]

    async def list(self, *, label: str | None = None, page_size: int = 100) -> AsyncIterator[Node]:
        """Yield every node (optionally filtered by label), auto-following X-Next-Cursor."""
        cursor: str | None = None
        prev_cursor: str | None = None
        while True:
            params: dict[str, Any] = {"limit": page_size}
            if label is not None:
                params["label"] = label
            if cursor is not None:
                params["cursor"] = cursor
            res = await self._t.request("GET", "/nodes", params=params)
            for d in res.data or []:
                yield Node.from_dict(d)
            cursor = res.headers.get("X-Next-Cursor")
            if not cursor or cursor == prev_cursor:
                return
            prev_cursor = cursor

File: clients/go/nodes.go
package graphdb

import (
	"context"
	"encoding/json"
	"fmt"
	"iter"
	"net/http"
	"net/url"
	"strconv"
)

// NodeInput is one node for BatchCreate.
type NodeInput struct {
	Labels     []string       `json:"labels"`
	Properties map[string]any `json:"properties"`
}

// ListOptions controls node listing.
type ListOptions struct {
	Label    string
	PageSize int // default 100 if <= 0
}

func (n *Nodes) Create(ctx context.Context, labels []string, props map[string]any) (*Node, error) {
	if labels == nil {
		labels = []string{}
	}
	if props == nil {
		props = map[string]any{}
	}
	res, err := n.t.request(ctx, http.MethodPost, "/nodes",
		map[string]any{"labels": labels, "properties": props}, nil)
	if err != nil {
		return nil, err
	}
	var out Node
	return &out, json.Unmarshal(res.data, &out)
}

func (n *Nodes) Get(ctx context.Context, id uint64) (*Node, error) {
	res, err := n.t.request(ctx, http.MethodGet, fmt.Sprintf("/nodes/%d", id), nil, nil)
	if err != nil {
		return nil, err
	}
	var out Node
	return &out, json.Unmarshal(res.data, &out)
}

func (n *Nodes) Update(ctx context.Context, id uint64, props map[string]any) (*Node, error) {
	res, err := n.t.request(ctx, http.MethodPut, fmt.Sprintf("/nodes/%d", id),
		map[string]any{"properties": props}, nil)
	if err != nil {
		return nil, err
	}
	var out Node
	return &out, json.Unmarshal(res.data, &out)
}

func (n *Nodes) Delete(ctx context.Context, id uint64) error {
	_, err := n.t.request(ctx, http.MethodDelete, fmt.Sprintf("/nodes/%d", id), nil, nil)
	return err
}

func (n *Nodes) BatchCreate(ctx context.Context, nodes []NodeInput) ([]Node, error) {
	res, err := n.t.request(ctx, http.MethodPost, "/nodes/batch",
		map[string]any{"nodes": nodes}, nil)
	if err != nil {
		return nil, err
	}
	var out struct {
		Nodes []Node `json:"nodes"`
	}
	return out.Nodes, json.Unmarshal(res.data, &out)
}

// List streams every node (optionally filtered by label), auto-following the
// X-Next-Cursor response header. On error the iterator yields one final
// (zero, err) pair and stops.
func (n *Nodes) List(ctx context.Context, opts ListOptions) iter.Seq2[Node, error] {
	pageSize := opts.PageSize
	if pageSize <= 0 {
		pageSize = 100
	}
	return func(yield func(Node, error) bool) {
		var cursor, prev string
		for {
			params := url.Values{}
			params.Set("limit", strconv.Itoa(pageSize))
			if opts.Label != "" {
				params.Set("label", opts.Label)
			}
			if cursor != "" {
				params.Set("cursor", cursor)
			}
			res, err := n.t.request(ctx, http.MethodGet, "/nodes", nil, params)
			if err != nil {
				yield(Node{}, err)
				return
			}
			var page []Node
			if err := json.Unmarshal(res.data, &page); err != nil {
				yield(Node{}, err)
				return
			}
			for _, nd := range page {
				if !yield(nd, nil) {
					return
				}
			}
			cursor = res.header.Get("X-Next-Cursor")
			if cursor == "" || cursor == prev {
				return
			}
			prev = cursor
		}
	}
}

// ListAll collects every node into a slice (convenience over List).
func (n *Nodes) ListAll(ctx context.Context, opts ListOptions) ([]Node, error) {
	var out []Node
	for nd, err := range n.List(ctx, opts) {
		if err != nil {
			return nil, err
		}
		out = append(out, nd)
	}
	return out, nil
}
